strawberryfields/apps/similarity.py

```diff
@@ -306,33 +320,148 @@ def event_cardinality(photon_number: int, max_count_per_mode: int, modes: int) -
+
+    **Example usage:**
+
+    >>> graph = nx.complete_graph(8)
+    >>> _prob_orbit_exact(graph, [2, 1, 1])
+    0.03744
+
+    Args:
+        graph (nx.Graph): input graph encoded in the GBS device
+        orbit (list[int]): orbit for which to calculate the probability
+        n_mean (float): total mean photon number of the GBS device
+        loss (float): fraction of photons lost in GBS
+
+    Returns:
+        float: exact orbit probability
+    """
+
+    if n_mean < 0:
+        raise ValueError("Mean photon number must be non-negative")
+    if not 0 <= loss <= 1:
+        raise ValueError("Loss parameter must take a value between zero and one")
+
+    modes = graph.order()
+    photons = sum(orbit)
+    A = nx.to_numpy_array(graph)
+    mean_photon_per_mode = n_mean / float(modes)
+
+    p = sf.Program(modes)
+
+    with p.context as q:
+        sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
+
+        if loss:
+            for _q in q:
+                sf.ops.LossChannel(1 - loss) | _q
+
+    eng = sf.LocalEngine(backend="gaussian")
+    result = eng.run(p)
+
+    click = orbit + [0] * (modes - len(orbit))
+    prob = 0
+
+    for pattern in multiset_permutations(click):
+        prob += result.state.fock_prob(pattern, cutoff=photons + 1)
+
+    return prob
+
+
+def _prob_event_exact(
+    graph: nx.Graph,
+    photon_number: int,
+    max_count_per_mode: int,
+    n_mean: float = 5,
+    loss: float = 0.0,
+) -> float:
+    r"""Gives the exact GBS probability of a given event according to the input graph.
+
+    Events are made up of multiple orbits. To calculate event probability, we can sum over
+    its constituent orbit probabilities using :func:`_prob_orbit_exact`.
+
+    **Example usage:**
+
+    >>> graph = nx.complete_graph(8)
+    >>> _prob_event_exact(graph, 4, 2)
+    0.11077
+
+    Args:
+        graph (nx.Graph): input graph encoded in the GBS device
+        photon_number (int): number of photons in the event
+        max_count_per_mode (int): maximum number of photons per mode in the event
+        n_mean (float): total mean photon number of the GBS device
+        loss (float): fraction of photons lost in GBS
+
+    Returns:
+        float: exact event probability
+    """
+
+    if n_mean < 0:
+        raise ValueError("Mean photon number must be non-negative")
+    if not 0 <= loss <= 1:
+        raise ValueError("Loss parameter must take a value between zero and one")
+    if photon_number < 0:
+        raise ValueError("Photon number must not be below zero")
+    if max_count_per_mode < 0:
+        raise ValueError("Maximum number of photons per mode must be non-negative")
+
+    prob = 0
+
+    for orbit in orbits(photon_number):
+        if max(orbit) <= max_count_per_mode:
+            prob += _prob_orbit_exact(graph, orbit, n_mean, loss)
+    return prob
+
+
+def _prob_orbit_mc(
+    graph: nx.Graph,
+    orbit: list,
+    n_mean: float = 5,
+    mc_samples: int = 1000,

```

Comment:
What is the motivation for changing the `mc_samples` argument name (before it was `samples`)?

Reply:
Note that if we have `prob_orbit_mc` be public (i.e. without the first underscore), then changing the argument name from `samples` to `mc_samples` could break some user's code, e.g., if before they had:
`prob_orbit_mc(g, orb, samples=10)`

Reply:
The motivation is that in the sampling-based feature vector functions, we use ``samples`` to represent the list of all pre-produced samples. I wanted to differentiate between the argument ``samples`` used in MC functions where it refers to the number of samples to be used in the approximation.

Reply:
I dont understand your second comment. Will there not be an error raised even if the function is public. ``sample`` is not a defined argument for ``prob_orbit_mc``.